registers/src/csrng.rs

```diff
@@ -4,16 +4,48 @@
 //
 #![allow(clippy::erasing_op)]
 #![allow(clippy::identity_op)]
+/// A zero-sized type that represents ownership of this
+/// peripheral, used to get access to a Register lock. Most
+/// programs create one of these in unsafe code near the top of
+/// main(), and pass it to the driver responsible for managing
+/// all access to the hardware.
+pub struct CsrngReg {
+    _priv: (),
+}
+impl CsrngReg {
+    pub const PTR: *mut u32 = 0x20002000 as *mut u32;
+    /// # Safety
+    ///
+    /// Caller must ensure that all concurrent use of this
+    /// peripheral in the firmware is done so in a compatible
+    /// way. The simplest way to enforce this is to only call
+    /// this function once.
+    pub unsafe fn new() -> Self {
+        Self { _priv: () }
+    }
+    /// Returns a register block that can be used to read
+    /// registers from this peripheral, but cannot write.
+    pub fn regs(&self) -> RegisterBlock<ureg::RealMmio> {
+        RegisterBlock {
+            ptr: Self::PTR,
+            mmio: core::default::Default::default(),
+        }
+    }
+    /// Return a register block that can be used to read and
+    /// write this peripheral's registers.
+    pub fn regs_mut(&mut self) -> RegisterBlock<ureg::RealMmioMut> {
+        RegisterBlock {
+            ptr: Self::PTR,
+            mmio: core::default::Default::default(),
+        }
+    }
+}
 #[derive(Clone, Copy)]
-pub struct RegisterBlock<TMmio: ureg::Mmio + core::borrow::Borrow<TMmio> = ureg::RealMmio> {
+pub struct RegisterBlock<TMmio: ureg::Mmio + core::borrow::Borrow<TMmio>> {
```

Comment:
So I get `CsrngReg` is exclusive, i.e., there's no way to have concurrent instances without dipping into `unsafe` multiple times. But what about the `RegisterBlock`s? They're `Copy`, which seems subversive to the lifetime tying `&self/&mut self` to `RealMmio/RealMmioMut`.

For example, this code rightly does _not_ compile:
```rust
fn test_multiple_register_blocks() {
    let mut csrng_reg = unsafe { CsrngReg::new() };
    let regs_mut1 = csrng_reg.regs_mut();
    let regs_mut2 = csrng_reg.regs_mut();
    println!("pop FIFO: {}", regs_mut2.genbits().read());
    println!("pop FIFO: {}", regs_mut1.genbits().read());
}
```
```
error[E0499]: cannot borrow `csrng_reg` as mutable more than once at a time
  --> drivers/test-fw/src/bin/csrng_tests.rs:93:21
   |
92 |     let regs_mut1 = csrng_reg.regs_mut();
   |                     -------------------- first mutable borrow occurs here
93 |     let regs_mut2 = csrng_reg.regs_mut();
   |                     ^^^^^^^^^^^^^^^^^^^^ second mutable borrow occurs here
94 |     println!("pop FIFO: {}", regs_mut2.genbits().read());
95 |     println!("pop FIFO: {}", regs_mut1.genbits().read());
   |                              ------------------- first borrow later used here

For more information about this error, try `rustc --explain E0499`.
```

But `Copy`ing `regs_mut1` compiles:
```rust
fn test_multiple_register_blocks() {
    let mut csrng_reg = unsafe { CsrngReg::new() };
    let regs_mut1 = csrng_reg.regs_mut();
    let regs_mut2 = regs_mut1;
    println!("pop FIFO: {}", regs_mut2.genbits().read());
    println!("pop FIFO: {}", regs_mut1.genbits().read());
}
```

Reply:
This is on purpose, as there are situations where you need multiple RegRef references alive from the same register block (for example, to pass to utility functions). Because the RegisterBlock has to expose all the registers as member functions, the RegRef has to take the lifetime of the entire RegisterBlock.

Fortunately, the lifetime of all the register blocks (and RegRefs) is tied to the csrng_reg instance, and if you try to store those somewhere outside of test_multiple_register_blocks(), it will fail to compile.

Reply:
To be clearer: RegisterBlock doesn't necessarily have to be Copy, but removing Copy doesn't make things much safer. Arrays and RegRefs all share the lifetime with RegisterBlock, where you can have multiple of them alive at once. Some of the drivers rely on this behavior, for example: https://github.com/chipsalliance/caliptra-sw/blob/main/drivers/src/kv_access.rs#L130

Reply:
The main thing I care about is that whoever holds the CsrngReg can decide who gets RegisterBlocks, and if they try to pass the CsrngReg to somewhere else, those RegisterBlocks won't be able to outlive the move.

Reply:
If Rust member functions ever support partial borrows in the future (where mbox.status() can return a different lifetime than mbox.fifo()), we can revisit this decision.